### PR TITLE
chore: Use cargo-udeps to check unused dependencies

### PR DESCRIPTION
**Pull request by tottoto**
## Motivation

Checks unused dependencies more strictly.

## Solution

Replaces cargo-machete with cargo-udeps, and fixes detected dependencies and features.

### CODE REVIEW

File: tonic/src/transport/service/discover.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ use std::{
 };
 use tokio::sync::mpsc::Receiver;
 
-use tokio_stream::Stream;
+use futures_core::Stream;
```
Comment (LucioFranco) [inline tonic/src/transport/service/discover.rs:12]: why this change?

Comment (tottoto) [reply tonic/src/transport/service/discover.rs:12]: As tonic does not use tokio-stream except Stream which is the re-exporting of futures-core, we could use futures-core directly to reduce dependencies. This is detected by cargo-udeps.

Comment (tottoto) [reply tonic/src/transport/service/discover.rs:12]: Can I ask the purpose of favoring `tokio-stream` with unused additional dependencies over `futures-core`?

Comment (LucioFranco) [reply tonic/src/transport/service/discover.rs:12]: Ideally, we'd like to move off of the futures crates (stream is complicated since we expected it to land in std by now) since they are not as maintained as the tokio ones. Its a bit of a complicated history.

Comment (tottoto) [reply tonic/src/transport/service/discover.rs:12]: Thanks for your answer. I am a bit confused, and would like to ask some more questions to better update this pull request. Even if we depend on `tokio-stream` instead of `futures-core`, as `Stream` we use is still provided by `futures-core`, we effectively depend on `futures-core`. Therefore I think that even if we use `tokio-stream` this time, we just add dependencies that we don't use and can't achieve our goal of not using `futures-core`. I would like to know a little more about this point.

Comment (LucioFranco) [reply tonic/src/transport/service/discover.rs:12]: Eventually, `Stream` will go into `std` so then we will only depend on `tokio-stream` but for now we can't do that since progress has been slow there. That said, I would rather be positioned for that future than live with one extra dependency right now if that makes sense. In addition, we control tokio-stream so its better to depend on that then directly depend on another package if that makes sense. Happy to explain more over discord/email/what ever if you're interested 😄

Comment (tottoto) [reply tonic/src/transport/service/discover.rs:12]: Thank you for your kind explanation!

> we control tokio-stream so its better to depend on that then directly depend on another package if that makes sense

Thanks to this point mainly, I think I reached to understand why we should depend on `tokio-steam` instead of `futures-core`. I will investigate whether it is possible to configure that we should explicitly depend on `tokio-stream`.

Comment (tottoto) [reply tonic/src/transport/service/discover.rs:12]: @LucioFranco Should we replace the existing all `futures_core::Stream` with `tokio_stream::Stream` in this pull request, or should it be done in the another pull request?

Comment (tottoto) [reply tonic/src/transport/service/discover.rs:12]: I found that the `cargo-udeps` checking will success by replacing all `futures_core::Stream` with `tokio_stream::Stream`! Therefore I think we should do it in this pull request.

**Comment by tottoto** [comment]
`cargo-deny` is failed due to duplicate version of `regex-syntax`. It will be fixed in #1373.

File: tonic/src/transport/service/discover.rs (hunk 1 above)
Comment (LucioFranco) [reply tonic/src/transport/service/discover.rs]: Yeah, we should convert everything from futures crates over. I think I originally used it but would like to remove it but I have not had time. Feel free to start on that!

**Comment by LucioFranco** [comment]
I updated your PR and with the commit from master that fixes it all. Should be good to go.

**Comment by tottoto** [comment]
The msrv test is failed because of https://github.com/hyperium/tonic/pull/1364/commits/9d9458d9e632613f35f34c877bc162b3a03967b6. `std::task::ready` (and `core::task::ready`) was stabilized at 1.64.0, but our msrv is 1.60.

**Review by LucioFranco — APPROVED** [review]
Sorry for the delay! Thanks!
